streams/src/main/java/org/apache/kafka/streams/query/InternalQueryResultUtil.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.query;
+
+import org.apache.kafka.common.annotation.InterfaceStability.Unstable;
+
+import java.util.List;
+
+/**
+ * This utility class is an internal API, but it must be located in the same package as
+ * {@link QueryResult} so that it can access the package-private constructor
+ * {@link QueryResult#QueryResult(Object, List, Position)}
+ */
+@Unstable
+public final class InternalQueryResultUtil {
+    // uninstantiable utility class
+    public InternalQueryResultUtil() {}
+
+    public static <R> QueryResult<R> copyAndSubstituteDeserializedResult(
+        final QueryResult<?> rawResult,
+        final R deserializedResult) {
```

Comment:
Not sure how many cases we have, but could we instead pass in the deserializer and all deserialization logic in this class? So far there are two case, plain value and Iterator case only.

Reply:
It seems like that would split the deserialization responsibility between the Metered stores (whose job it is to handle de/serialization) and the QueryResult domain, whose job it is to just carry results, so I reckon it's better to keep things as they are now.

I agree with your motivation to clean up the Metered stores' logic, though. I have some ideas about that:
* https://github.com/apache/kafka/pull/11614
* https://github.com/vvcephei/kafka/pull/1

Those PRs are sketches of possible approaches to https://issues.apache.org/jira/browse/KAFKA-13526

Reply:
Ack. Makes sense.